raysect/core/math/quaternion.pyx

```diff
@@ -0,0 +1,556 @@
+
+            return q1.mul_scalar(s)
+
+        elif isinstance(x, Quaternion) and isinstance(y, Quaternion):
+
+            q1 = <Quaternion> x
+            q2 = <Quaternion> y
+
+            return q1.mul(q2)
+
+        else:
+
+            return NotImplemented()
+
+    @cython.cdivision(True)
+    def __truediv__(object x, object y):
+        """Division operator.
+
+        .. code-block:: pycon
+
+            >>> Quaternion(1, 0, 1, 0) / 2
+            Quaternion(0.5, <0.0, 0.5, 0.0>)
+            >>> Quaternion(1, 0, 1, 0) / Quaternion(0, 1, 2, 3)
+            Quaternion(0.14286, <-0.28571, -0.14286, -0.14286>)
+        """
+
+        cdef double d
+        cdef Quaternion q1, q2, q2_inv
+
+        if isinstance(x, Quaternion) and isinstance(y, numbers.Real):
+
+            d = <double> y
+            q1 = <Quaternion> x
+
+            return q1.div_scalar(d)
+
+        elif isinstance(x, Quaternion) and isinstance(y, Quaternion):
+
+            q1 = <Quaternion> x
+            q2 = <Quaternion> y
+
+            return q1.div(q2)
+
+        else:
+
+            raise TypeError("Unsupported operand type. Expects a real number.")
+
+    cdef Quaternion neg(self):
+        """
+        Fast negation operator.
+
+        This is a cython only function and is substantially faster than a call
+        to the equivalent python operator.
+        """
+
+        return new_quaternion(-self.s, -self.x, -self.y, -self.z)
+
+    cdef Quaternion add(self, Quaternion q2):
+        """
+        Fast addition operator.
+
+        This is a cython only function and is substantially faster than a call
+        to the equivalent python operator.
+        """
+
+        cdef Quaternion q1 = self
+
+        return new_quaternion(q1.s + q2.s, q1.x + q2.x, q1.y + q2.y, q1.z + q2.z)
+
+    cdef Quaternion sub(self, Quaternion q2):
+        """
+        Fast subtraction operator.
+
+        This is a cython only function and is substantially faster than a call
+        to the equivalent python operator.
+        """
+
+        cdef Quaternion q1 = self
+
+        return new_quaternion(q1.s - q2.s, q1.x - q2.x, q1.y - q2.y, q1.z - q2.z)
+
+    cdef Quaternion mul(self, Quaternion q2):
+        """
+        Fast multiplication operator.
+
+        This is a cython only function and is substantially faster than a call
+        to the equivalent python operator.
+        """
+
+        cdef Quaternion q1 = self
+        cdef double ns, nx, ny, nz
+
+        ns = q1.s*q2.s - q1.x*q2.x - q1.y*q2.y - q1.z*q2.z
+        nx = q1.s*q2.x + q1.x*q2.s + q1.y*q2.z - q1.z*q2.y
+        ny = q1.s*q2.y - q1.x*q2.z + q1.y*q2.s + q1.z*q2.x
+        nz = q1.s*q2.z + q1.x*q2.y - q1.y*q2.x + q1.z*q2.s
+
+        return new_quaternion(ns, nx, ny, nz)
+
+    cdef Quaternion mul_scalar(self, double d):
+
+        cdef Quaternion q = self
+
+        return new_quaternion(d * q.s, d * q.x, d * q.y, d * q.z)
+
+    @cython.cdivision(True)
+    cpdef Quaternion inv(self):
```

Comment:
Please use inverse() not inv(), as per AffineMatrix3D: cpdef AffineMatrix3D inverse(self)

Reply:
Resolved in a4d9d8a